### PR TITLE
Setup Bors

### LINKED ISSUE #124: Setup Bors PR Merge Bot
### Description

We'll probably want to use Bors for merging pull requests. It's simple to setup and I can do it if we want to.

The goal is to make sure that everything that gets merged to master has actually passed CI in it's exact state. As it stands, if we merge multiple pull requests off of master, even if each branch passed CI, we don't know that the _combination_ of the pull requests will pass CI once they are merged to master.

I'd like to hear thoughts form @odecay and @erlend-sh if you guys have any input.

### Alternatives & Prior Art

See this comment for background: https://github.com/fishfight/punchy/pull/111#pullrequestreview-1048578729.

### PR DESCRIPTION
**Pull request by 64kramsystem**
Closes #124.

### CODE REVIEW

**Comment by 64kramsystem** [comment]
This can be tested by running the bors comment (see help; I don't want to trigger it unintentionally).

The documentation is very vague (e.g. about the location of the configfile; it also refers to workflow names, but they're actually the workflow **file**names), so this may potentially take a few attempts.)

File: bors.toml
Comment (zicklag) [inline bors.toml:1]: I prefer to add the `use-squash-merge` option which makes the commit history look like it was done by normal people instead of the bors bot. See the [Skip'n Go commit history](https://github.com/katharostech/skipngo/commits/master) for an example of commits merged with that option.

```suggestion
status = ["ci"]
use_squash_merge = true
```

**Review by zicklag — CHANGES_REQUESTED** [review]

**Comment by zicklag** [comment]
I put the `bors.toml` in the `.github` directory in my project, though it might also work in the root dir I think.

File: bors.toml
Comment (64kramsystem) [reply bors.toml:1]: Squash merges are going to void the advantages of granular commits.

Granular commits allow a more precise bisecting to find bugs, and make understanding steps to implement functionalities easier.

It already happened that I wasn't able to fully understand how to implement a functionality because when I looked at the commit, it was a wall of code.

AFAIK Github preserves the individual commits, but not in the local repository. This means that local tools (IDEs/CLI) won't be able to dig in the history.

**Comment by 64kramsystem** [comment]
> I put the `bors.toml` in the `.github` directory in my project, though it might also work in the root dir I think.

Ah, I'll move it. This is something that the documentation doesn't specify :roll_eyes: .

File: bors.toml
Comment (zicklag) [reply bors.toml]: Valid point. Go ahead and leave it out then.

**Comment by odecay** [comment]
Don't know much about this but from what you all have said it seems like bors will be helpful, looking forward to getting to adapt to workflow to whatever you decide. :+1:

**Review by odecay — APPROVED** [review]

**Comment by 64kramsystem** [comment]
I'm going to try and merge it via bors command; I've addressed one of the @zicklag review note, and the other one has been dropped after discussion.

**Comment by 64kramsystem** [comment]
woohooo! it seems it's working :star_struck:

**Comment by zicklag** [comment]
Bors timed out because we need to setup our CI pipeline to run on  the `staging` and `trying` branches that bors will use to do merged commits.

bors retry

**Review by zicklag — APPROVED** [review]

**Comment by 64kramsystem** [comment]
> Bors timed out because we need to setup our CI pipeline to run on the `staging` and `trying` branches that bors will use to do merged commits.
> 
> bors retry

Hmmm... still timed out :thinking: Any idea why? I don't have experience with it :scream:

**Comment by zicklag** [comment]
I think I figured it.

Bors merge

**Comment by zicklag** [comment]
bors help

**Comment by zicklag** [comment]
bors ping

**Comment by 64kramsystem** [comment]
LOL! :joy:

**Comment by zicklag** [comment]
bors merge

Please?

**Comment by zicklag** [comment]
🎉 Now we're in business! :)

**Comment by odecay** [comment]
What is the basic workflow we intend to use now with bors?

**Comment by zicklag** [comment]
When somebody opens a pull request, you review it just like normal. Once it's ready to merge instead of manually merging it you just say `bors merge` in a comment, on it's own line. Then bors will grab the PR, create a merge commit, and run CI on that merge commit. If CI passes on the merge commit, then it merges to master and closes the PR automatically, so you don't have to wait for CI to pass.

**Comment by odecay** [comment]
Sounds good.

**Pull request merged**
